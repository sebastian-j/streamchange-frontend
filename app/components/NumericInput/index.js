import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Container = styled.div`
  display: flex;
  flex-direction: row;
`;

const Label = styled.div`
  color: ${props => props.theme.staticTextColor};
  line-height: 30px;
  margin-right: 5px;
`;
const Button = styled.button`
  background: ${props => props.theme.buttonBackground};
  border: 1px solid ${props => props.theme.color};
  border-radius: 0 4px 4px 0;
  color: ${props => props.theme.buttonTextColor};
  font-size: 1.5rem;
  line-height: 23px;
  height: 30px;
  width: 30px;
  text-decoration: none;
  outline: none;
  ${({ left }) =>
    left &&
    `
    border-radius: 4px 0 0 4px;
  `}
  &:hover {
    background-color: ${props => props.theme.buttonBackgroundHover};
    color: ${props => props.theme.buttonTextColorHover};
  }
`;

const NumberDisplay = styled.input`
  background: transparent;
  border: 1px solid gray;
  color: ${props => props.theme.staticTextColor};
  font-size: 1.5rem;
  height: 100%;
  max-width: 70px;
  padding: 0 10px;
`;

const NumericInput = props => {
  const inc = () => {
    let val = props.value;
    if (props.maxValue === null || val + props.step < props.maxValue) {
      val = props.value + props.step;
    } else {
      val = props.maxValue;
    }
    if (val !== props.value) props.onChange(Number(val));
  };
  const dec = () => {
    let val = props.value;
    if (props.minValue === null || val - props.step > props.minValue) {
      val = props.value - props.step;
    } else {
      val = props.minValue;
    }
    if (val !== props.value) props.onChange(Number(val));
  };
  const checkValue = () => {
    let val = props.value;
    if (val < props.minValue) val = props.minValue;
    if (val > props.maxValue) val = props.maxValue;
    if (Number.isNaN(Number(val))) val = props.value;
    props.onChange(Number(val));
  };
  return (
    <Container>
      <Label>{props.label}</Label>
      <Button left="true" onClick={dec} type="button">
        -
      </Button>
      <NumberDisplay
        value={props.value}
        onChange={event => props.onChange(event.target.value)}
        onBlur={checkValue}
      />
      <Button onClick={inc} type="button">
        +
      </Button>
    </Container>
  );
};

NumericInput.propTypes = {
  label: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.number,
  minValue: PropTypes.number,
  maxValue: PropTypes.number,
  step: PropTypes.number,
};
NumericInput.defaultProps = {
  value: 0,
  minValue: null,
  maxValue: null,
  step: 1,
};

export default NumericInput;
