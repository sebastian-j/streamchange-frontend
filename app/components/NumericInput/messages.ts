import { defineMessages } from 'react-intl';

export const scope = 'streamchange.components.NumericInput';

export default defineMessages({
  decreaseButton: {
    id: `${scope}.decrease`,
    defaultMessage: 'Decrease by one',
  },
  increaseButton: {
    id: `${scope}.increase`,
    defaultMessage: 'Increase by one',
  },
});
