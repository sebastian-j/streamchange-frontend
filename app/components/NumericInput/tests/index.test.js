import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import NumericInput from '../index';

const renderer = new ShallowRenderer();

describe('<NumericInput />', () => {
  it('should render and match the snapshot', () => {
    renderer.render(<NumericInput onChange={() => 0} />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
