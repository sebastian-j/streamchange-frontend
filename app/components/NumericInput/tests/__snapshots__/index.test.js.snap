// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumericInput /> should render and match the snapshot 1`] = `
<ForwardRef>
  <ForwardRef />
  <ForwardRef
    left="true"
    onClick={[Function]}
    type="button"
  >
    -
  </ForwardRef>
  <ForwardRef>
    0
  </ForwardRef>
  <ForwardRef
    onClick={[Function]}
    type="button"
  >
    +
  </ForwardRef>
</ForwardRef>
`;
